File: typescript/tests/payment/application/commands/paymentCommands.test.ts
import { describe, expect, it } from 'vitest';
import { generateId } from '@/common/domain/identity/entityId.js';
import { ClientReference } from '@/common/domain/primitive/clientReference.js';
import { Description } from '@/common/domain/primitive/description.js';
import { Money } from '@/common/domain/primitive/money.js';
import { Timestamp } from '@/common/domain/primitive/timestamp.js';
import { acceptTransaction } from '@/payment/application/commands/acceptTransaction.js';
import { expirePayment } from '@/payment/application/commands/expirePayment.js';
import { refundTransaction } from '@/payment/application/commands/refundTransaction.js';
import { rejectTransaction } from '@/payment/application/commands/rejectTransaction.js';
import { requestPayment } from '@/payment/application/commands/requestPayment.js';
import { startTransaction } from '@/payment/application/commands/startTransaction.js';
import { PaymentId } from '@/payment/domain/payment/paymentId.js';
import { Provider } from '@/payment/domain/payment/provider.js';
import { ProviderReference } from '@/payment/domain/payment/providerReference.js';
import { TransactionId } from '@/payment/domain/payment/transactionId.js';
import { Uuid } from '@/common/domain/primitive/uuid.js';

const A_CLIENT_REFERENCE = new ClientReference(Uuid.fromString('00000000-0000-0000-0000-000000000001'));

const REQUESTED_AT = new Timestamp(new Date('2026-06-07T10:00:00.000Z'));

describe('PaymentCommands', () => {
  describe('RequestPayment validation', () => {
    it('should fail if parameters are null', () => {
      const paymentId = generateId((value) => new PaymentId(value));
      const amount = new Money(10);

      expect(() => requestPayment(null as unknown as PaymentId, A_CLIENT_REFERENCE, amount, REQUESTED_AT)).toThrow();
      expect(() => requestPayment(paymentId, null as unknown as ClientReference, amount, REQUESTED_AT)).toThrow();
      expect(() => requestPayment(paymentId, A_CLIENT_REFERENCE, null as unknown as Money, REQUESTED_AT)).toThrow();
      expect(() => requestPayment(paymentId, A_CLIENT_REFERENCE, amount, null as unknown as Timestamp)).toThrow();
    });
  });

  describe('StartTransaction validation', () => {
    it('should fail if parameters are null', () => {
      const paymentId = generateId((value) => new PaymentId(value));
      const providerReference = new ProviderReference(Uuid.generate());
      const amount = new Money(10);
      const startedAt = REQUESTED_AT;

      expect(() => startTransaction(null as unknown as PaymentId, Provider.GIFT_CARD, providerReference, amount, startedAt)).toThrow();
      expect(() => startTransaction(paymentId, null as unknown as Provider, providerReference, amount, startedAt)).toThrow();
      expect(() => startTransaction(paymentId, Provider.GIFT_CARD, providerReference, null as unknown as Money, startedAt)).toThrow();
      expect(() => startTransaction(paymentId, Provider.GIFT_CARD, providerReference, amount, null as unknown as Timestamp)).toThrow();
    });
  });

  describe('RefundTransaction validation', () => {
    it('should fail if parameters are null', () => {
      const paymentId = generateId((value) => new PaymentId(value));
      const amount = new Money(10);

      expect(() => refundTransaction(null as unknown as PaymentId, amount)).toThrow();
      expect(() => refundTransaction(paymentId, null as unknown as Money)).toThrow();
    });
  });

  describe('RejectTransaction validation', () => {
    it('should fail if parameters are null or blank', () => {
      const paymentId = generateId((value) => new PaymentId(value));
      const transactionId = generateId((value) => new TransactionId(value));

      expect(() => rejectTransaction(null as unknown as PaymentId, transactionId, new Description('reason'))).toThrow();
      expect(() => rejectTransaction(paymentId, null as unknown as TransactionId, new Description('reason'))).toThrow();
      expect(() => rejectTransaction(paymentId, transactionId, null as unknown as Description)).toThrow();
      expect(() => rejectTransaction(paymentId, transactionId, new Description(''))).toThrow();
    });
  });

  describe('AcceptTransaction validation', () => {
    it('should fail if parameters are null', () => {
      const paymentId = generateId((value) => new PaymentId(value));
      const transactionId = generateId((value) => new TransactionId(value));
      const providerCompletedAt = REQUESTED_AT;

      expect(() => acceptTransaction(null as unknown as PaymentId, transactionId, providerCompletedAt)).toThrow();
      expect(() => acceptTransaction(paymentId, null as unknown as TransactionId, providerCompletedAt)).toThrow();
      expect(() => acceptTransaction(paymentId, transactionId, null as unknown as Timestamp)).toThrow();
    });
  });

  describe('ExpirePayment validation', () => {
    it('should fail if aggregate id is null', () => {
      expect(() => expirePayment(null as unknown as PaymentId)).toThrow();
    });
  });
});

File: java/src/test/java/it/giannibombelli/wsc2026/payment/application/commands/PaymentCommandsTest.java
package it.giannibombelli.wsc2026.payment.application.commands;

import it.giannibombelli.wsc2026.common.domain.identity.EntityId;
import it.giannibombelli.wsc2026.common.domain.primitive.ClientReference;
import it.giannibombelli.wsc2026.common.domain.primitive.Description;
import it.giannibombelli.wsc2026.common.domain.primitive.Money;
import it.giannibombelli.wsc2026.common.domain.primitive.Timestamp;
import it.giannibombelli.wsc2026.payment.domain.payment.PaymentId;
import it.giannibombelli.wsc2026.payment.domain.payment.Provider;
import it.giannibombelli.wsc2026.payment.domain.payment.ProviderReference;
import it.giannibombelli.wsc2026.payment.domain.payment.TransactionId;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;

import java.math.BigDecimal;
import java.time.Instant;
import java.util.UUID;

import static org.assertj.core.api.Assertions.assertThatThrownBy;

class PaymentCommandsTest {

    @Nested
    class RequestPaymentValidation {
        @Test
        void shouldFailIfParametersAreNullOrBlank() {
            PaymentId paymentId = EntityId.generate(PaymentId::new);
            Money amount = new Money(BigDecimal.TEN);
            Timestamp requestedAt = new Timestamp(Instant.parse("2026-06-07T10:00:00Z"));

            assertThatThrownBy(() -> new RequestPayment(null, new ClientReference(UUID.fromString("00000000-0000-0000-0000-000000000001")), amount, requestedAt))
                .isInstanceOf(IllegalArgumentException.class);

            assertThatThrownBy(() -> new RequestPayment(paymentId, null, amount, requestedAt))
                .isInstanceOf(IllegalArgumentException.class);

            assertThatThrownBy(() -> new RequestPayment(paymentId, new ClientReference((UUID) null), amount, requestedAt))
                .isInstanceOf(IllegalArgumentException.class);

            assertThatThrownBy(() -> new RequestPayment(paymentId, new ClientReference(UUID.fromString("00000000-0000-0000-0000-000000000001")), null, requestedAt))
                .isInstanceOf(IllegalArgumentException.class);

            assertThatThrownBy(() -> new RequestPayment(paymentId, new ClientReference(UUID.fromString("00000000-0000-0000-0000-000000000001")), amount, null))
                .isInstanceOf(IllegalArgumentException.class);
        }
    }

    @Nested
    class StartTransactionValidation {
        @Test
        void shouldFailIfParametersAreNull() {
            PaymentId paymentId = EntityId.generate(PaymentId::new);
            UUID providerReference = UUID.randomUUID();
            Money amount = new Money(BigDecimal.TEN);
            Timestamp startedAt = new Timestamp(Instant.parse("2026-06-07T10:00:00Z"));
            ProviderReference providerReferenceVo = new ProviderReference(providerReference);

            assertThatThrownBy(() -> new StartTransaction(null, Provider.GIFT_CARD, providerReferenceVo, amount, startedAt))
                .isInstanceOf(IllegalArgumentException.class);

            assertThatThrownBy(() -> new StartTransaction(paymentId, null, providerReferenceVo, amount, startedAt))
                .isInstanceOf(IllegalArgumentException.class);

            assertThatThrownBy(() -> new StartTransaction(paymentId, Provider.GIFT_CARD, providerReferenceVo, null, startedAt))
                .isInstanceOf(IllegalArgumentException.class);

            assertThatThrownBy(() -> new StartTransaction(paymentId, Provider.GIFT_CARD, providerReferenceVo, amount, null))
                .isInstanceOf(IllegalArgumentException.class);
        }
    }

    @Nested
    class RefundTransactionValidation {
        @Test
        void shouldFailIfParametersAreNull() {
            PaymentId paymentId = EntityId.generate(PaymentId::new);
            Money amount = new Money(BigDecimal.TEN);

            assertThatThrownBy(() -> new RefundTransaction(null, amount))
                .isInstanceOf(IllegalArgumentException.class);

            assertThatThrownBy(() -> new RefundTransaction(paymentId, null))
                .isInstanceOf(IllegalArgumentException.class);
        }
    }

    @Nested
    class RejectTransactionValidation {
        @Test
        void shouldFailIfParametersAreNullOrBlank() {
            PaymentId paymentId = EntityId.generate(PaymentId::new);
            TransactionId transactionId = EntityId.generate(TransactionId::new);

            assertThatThrownBy(() -> new RejectTransaction(null, transactionId, new Description("reason")))
                .isInstanceOf(IllegalArgumentException.class);

            assertThatThrownBy(() -> new RejectTransaction(paymentId, null, new Description("reason")))
                .isInstanceOf(IllegalArgumentException.class);

            assertThatThrownBy(() -> new RejectTransaction(paymentId, transactionId, null))
                .isInstanceOf(IllegalArgumentException.class);

            assertThatThrownBy(() -> new RejectTransaction(paymentId, transactionId, new Description("")))
                .isInstanceOf(IllegalArgumentException.class);
        }
    }

    @Nested
    class AcceptTransactionValidation {
        @Test
        void shouldFailIfParametersAreNull() {
            PaymentId paymentId = EntityId.generate(PaymentId::new);
            TransactionId transactionId = EntityId.generate(TransactionId::new);
            Timestamp providerCompletedAt = new Timestamp(Instant.parse("2026-06-07T10:00:00Z"));

            assertThatThrownBy(() -> new AcceptTransaction(null, transactionId, providerCompletedAt))
                .isInstanceOf(IllegalArgumentException.class);

            assertThatThrownBy(() -> new AcceptTransaction(paymentId, null, providerCompletedAt))
                .isInstanceOf(IllegalArgumentException.class);

            assertThatThrownBy(() -> new AcceptTransaction(paymentId, transactionId, null))
                .isInstanceOf(IllegalArgumentException.class);
        }
    }

    @Nested
    class ExpirePaymentValidation {
        @Test
        void shouldFailIfAggregateIdIsNull() {
            assertThatThrownBy(() -> new ExpirePayment(null))
                .isInstanceOf(IllegalArgumentException.class);
        }
    }
}
